crates/cubecl-reduce/src/launch.rs

```diff
@@ -221,20 +330,128 @@ pub fn reduce_slice_plane<N: Numeric, Instr: Reduce<N>>(
 
-        Instr::reduce(&mut accumulator, item, coordinates, true);
+        let reduction = Inst::reduce(&accumulator, item, coordinates, true);
+        Inst::update_accumulator(&mut accumulator, &reduction);
         let plane_dim = CUBE_DIM_X;
         first_index += plane_dim * stride;
         first_coordinate += plane_dim;
     }
     accumulator
 }
 
+#[cube]
+pub fn reduce_slice_shared<N: Numeric, Inst: Reduce<N>>(
+    items: Slice<Line<N>>,
+    start: u32,
+    end: u32,
+    stride: u32,
+    #[comptime] accumulator_size: u32,
+    #[comptime] line_size: u32,
+    #[comptime] line_mode: LineMode,
+) -> Inst::SharedAccumulator {
+    let mut accumulator = Inst::SharedAccumulator::allocate(accumulator_size, line_size);
+    Inst::SharedAccumulator::write(
+        &mut accumulator,
+        UNIT_POS,
+        Inst::null_accumulator(line_size),
+    );
+
+    let mut first_index = start;
+    let mut first_coordinate = 0;
+    while first_index < end {
+        let coordinate = first_coordinate + UNIT_POS;
+        let line_coordinate = match comptime!(line_mode) {
+            LineMode::Parallel => {
+                let mut coordinates = Line::empty(line_size).fill(coordinate * line_size);
+                if line_size > 1 {
+                    #[unroll]
+                    for j in 0..line_size {
+                        coordinates[j] += j;
+                    }
+                }
+                coordinates
+            }
+            LineMode::Perpendicular => Line::empty(line_size).fill(coordinate),
+        };
+        let index = first_index + UNIT_POS * stride;
+        let item = select(index < end, items[index], Inst::null_input(line_size));
+        let reduction = Inst::reduce(
+            &Inst::SharedAccumulator::read(&accumulator, UNIT_POS),
+            item,
+            line_coordinate,
+            false,
+        );
+        Inst::SharedAccumulator::write(&mut accumulator, UNIT_POS, reduction);
+        first_index += stride * CUBE_DIM;
+        first_coordinate += CUBE_DIM;
+    }
+    accumulator
+}
+
+/// Use all units within a cube to fuse an accumulator inplace like this with some padding if CUBE_DIM is not a power of 2.
+///
+/// ```ignored
+///
+///     0   1   2   3   4   5   6   7
+///     |   |   |   |   |   |   |   |
+///     +---+   +---+   +---+   +---+
+///     |       |       |       |
+///     +-------+       +-------+
+///     |               |
+///     +---------------+
+///     |
+///     *
+///
+/// ```
+///
+/// The outcome is stored in the first element of the accumulator and also returned by this function for convenience.
+///
+/// This assumes that the size of accumulator is exactly CUBE_DIM and leads to undefined behavior if not.
+#[cube]
+pub fn reduce_tree<N: Numeric, Inst: Reduce<N>>(
+    accumulator: &mut Inst::SharedAccumulator,
+) -> Inst::AccumulatorItem {
+    if comptime!(CUBE_DIM.is_power_of_two()) {
```

Comment:
The if and else clause have gotten quite similar, do you think we could refactor to mitigate duplication?